### PR TITLE
rp2pio: Support PIOv1 (rp2350) features

### PR DESCRIPTION
**Pull request by jepler**
This is needed to fully support the .mov_status and .fifo directives, as well as checking for PIO version compatibility between program & runtime hardware.

- [ ] does consider_instruction need to be extended
- [x] test on HW & test new instructions & new HW features

This works in tandem with https://github.com/adafruit/Adafruit_CircuitPython_PIOASM/pull/68

### CODE REVIEW

**Comment by jepler** [comment]
Going to punt on extending consider_instruction for now.

File: ports/raspberrypi/bindings/rp2pio/StateMachine.c [hunk 1]
```diff
@@ -841,6 +911,32 @@ MP_DEFINE_CONST_FUN_OBJ_1(rp2pio_statemachine_get_pc_obj, rp2pio_statemachine_ob
 MP_PROPERTY_GETTER(rp2pio_statemachine_pc_obj,
     (mp_obj_t)&rp2pio_statemachine_get_pc_obj);
 
+//|     rxfifo: memorymap.AddressRange
+//|     """Accecss the state machine's rxfifo directly
```
Comment (tannewt) [inline ports/raspberrypi/bindings/rp2pio/StateMachine.c:915]:
```suggestion
//|     """Access the state machine's rxfifo directly
```

File: ports/raspberrypi/bindings/rp2pio/StateMachine.c [hunk 2]
```diff
@@ -841,6 +911,32 @@ MP_DEFINE_CONST_FUN_OBJ_1(rp2pio_statemachine_get_pc_obj, rp2pio_statemachine_ob
...
+//|
+//|     If the state machine's fifo mode is ``txput`` then accessing this object
+//|     reads values stored by the ``mov rxfifo[], isr`` PIO instruction, and the
+//|     result of modifying it is undefined.
+//|
+//|     If the state machine's fifo mode is ``txget`` then modifying this object
+//|     writes values accessed by the ``mov osr, rxfifo[]`` PIO instruction, and
+//|     the result of accessing it is undefined..
```
Comment (tannewt) [inline ports/raspberrypi/bindings/rp2pio/StateMachine.c:923]:
```suggestion
//|     the result of accessing it is undefined.
```

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.c [hunk 3]
```diff
@@ -343,15 +346,34 @@ bool rp2pio_statemachine_construct(rp2pio_statemachine_obj_t *self,
...
+    enum pio_fifo_join join =
+        fifo_type != PIO_FIFO_JOIN_AUTO ? fifo_type
+        : !rx_fifo ? PIO_FIFO_JOIN_TX
+        : !tx_fifo ? PIO_FIFO_JOIN_RX
+        : PIO_FIFO_JOIN_NONE;
+    self->fifo_depth = (join == PIO_FIFO_JOIN_TX || join == PIO_FIFO_JOIN_RX) ? 8
+        #if PICO_PIO_VERSION > 0
+        : (join == PIO_FIFO_JOIN_PUTGET) ? 0
+        #endif
+        : 4;
```
Comment (tannewt) [inline ports/raspberrypi/common-hal/rp2pio/StateMachine.c:367]: Please add parens or switch to if statements. Nested ternary is very hard for me to read.

File: ports/raspberrypi/common-hal/rp2pio/StateMachine.h [hunk 4]
```diff
@@ -47,6 +51,7 @@ typedef struct {
     sm_buf_info current, once, loop;
     int background_stride_in_bytes;
     bool dma_completed, byteswap;
+    memorymap_addressrange_obj_t rxfifo_obj;
```
Comment (tannewt) [inline ports/raspberrypi/common-hal/rp2pio/StateMachine.h:54]: Want to wrap this in the version test to save object size on rp2040?

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for adding this support! Just a few minor things.

**Review by tannewt — APPROVED** [review]
Thank you! Looks good!

**Pull request merged**